collector/dfsr.go

```diff
@@ -0,0 +1,868 @@
+	FolderFilesReceivedTotal                       *prometheus.Desc
+	FolderRDCBytesReceivedTotal                    *prometheus.Desc
+	FolderRDCCompressedSizeOfFilesReceivedTotal    *prometheus.Desc
+	FolderRDCNumberofFilesReceivedTotal            *prometheus.Desc
+	FolderRDCSizeOfFilesReceivedTotal              *prometheus.Desc
+	FolderSizeOfFilesReceivedTotal                 *prometheus.Desc
+	FolderStagingSpaceInUse                        *prometheus.Desc
+	FolderStagingBytesCleanedUpTotal               *prometheus.Desc
+	FolderStagingBytesGeneratedTotal               *prometheus.Desc
+	FolderStagingFilesCleanedUpTotal               *prometheus.Desc
+	FolderStagingFilesGeneratedTotal               *prometheus.Desc
+	FolderUpdatesDroppedTotal                      *prometheus.Desc
+
+	// Volume source
+	VolumeDatabaseLookupsTotal           *prometheus.Desc
+	VolumeDatabaseCommitsTotal           *prometheus.Desc
+	VolumeUSNJournalUnreadPercentage     *prometheus.Desc
+	VolumeUSNJournalRecordsAcceptedTotal *prometheus.Desc
+	VolumeUSNJournalRecordsReadTotal     *prometheus.Desc
+
+	// Map of child collector functions used during collection
+	dfsrChildCollectors dfsrCollectorMap
+	// Internal counter for number of child collector failures during collection
+	dfsrChildCollectorFailure int
+}
+
+type dfsrCollectorMap map[string]dfsrCollectorFunc
+
+type dfsrCollectorFunc func(ctx *ScrapeContext, ch chan<- prometheus.Metric) (*prometheus.Desc, error)
+
+// Map Perflib sources to DFSR collector names
+// E.G. volume -> DFS Replication Service Volumes
+func dfsrGetPerfObjectName(collector string) string {
+	prefix := "DFS "
+	suffix := ""
+	switch collector {
+	case "connection":
+		suffix = "Replication Connections"
+	case "folder":
+		suffix = "Replicated Folders"
+	case "volume":
+		suffix = "Replication Service Volumes"
+	}
+	return (prefix + suffix)
+}
+
+// NewDFSRCollector is registered
+func NewDFSRCollector() (Collector, error) {
+	const subsystem = "dfsr"
+
+	enabled := expandEnabledChildCollectors(*dfsrEnabledCollectors)
+	perfCounters := make([]string, 0, len(enabled))
+	for _, c := range enabled {
+		perfCounters = append(perfCounters, dfsrGetPerfObjectName(c))
+	}
+	addPerfCounterDependencies(subsystem, perfCounters)
+
+	dfsrCollector := DFSRCollector{
+		// meta
+		dfsrScrapeDurationDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "collector_duration_seconds"),
+			"windows_exporter: Duration of an dfsr child collection.",
+			[]string{"collector"},
+			nil,
+		),
+		dfsrScrapeSuccessDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "collector_success"),
+			"windows_exporter: Whether a dfsr child collector was successful.",
+			[]string{"collector"},
+			nil,
+		),
+
+		// Connection
+		ConnectionBandwidthSavingsUsingDFSReplicationTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "bandwidth_savings_using_dfs_replication_total"),
```

Comment:
Bit of a nitpick, but we should have the unit in the name, as a suffix for gauges or just before the _total suffix for counters.
Goes for a lot of the names below as well, unfortunately.